contrib/babelfishpg_tds/src/backend/tds/guc.c

```diff
@@ -113,6 +113,17 @@ TdsGucDefaultPacketSizeCheck(int *newvalue, void **extra, GucSource source)
 	return true;
 }
 
+static void
+set_newval_to_product_version(char **newval)
+{
+	free(*newval);
+	*newval = strdup(product_version);
```

Comment:
Can we use `pstrdup` because it already contains OOM handling?

Reply:
`pstrdup` cannot be used here because it allocates the memory with `MemoryContextAlloc`, not with libc's `malloc`. And the initial memory of a `*newal` pointer is allocated in `parse_and_validate_value` with `guc_strdup` that [uses strdup under the hood](https://github.com/babelfish-for-postgresql/postgresql_modified_for_babelfish/blob/ff43eb60c73a7da7d20ea7d691edbc4b67ce0445/src/backend/utils/misc/guc.c#L5251). The value we assign to `*newval` is deallocated by `free` call in [AlterSystemSetConfigFile](https://github.com/babelfish-for-postgresql/postgresql_modified_for_babelfish/blob/ff43eb60c73a7da7d20ea7d691edbc4b67ce0445/src/backend/utils/misc/guc.c#L8988). So the proposed `set_newval_to_product_version` uses `strdup` to keep the logic exactly the same as now and to avoid the leak.

At the same time the motivation behind this assignment to `*newval` is not clear to me - the assigned value does not seem to be ever read. The value passed to `call_string_check_hook` in `*newval` pointer is a copy of the actual config value and this copy is deallocated immediately after the `call_string_check_hook` in both cases, when the hook returns `true` or `false`. Would it be better to remove this assignment completely by just deleting `*newval = product_version;` lines?

Reply:
@staticlibs thanks for your explanation! I agree with the first paragraph.

Regarding `*newval = product_version`, we need this because we return `true` in `check_version_number()` function. If we return `false` here, it can throw panic-level error and may cause DB restart.